### DISCUSSION TITLE
Automating workflows using GitHub Actions

### DISCUSSION THREAD

**Discussion by TauqeerAhmad5201** [community/community#36419] [Discover: GitHub Best Practices]
I am a Computer Science undergrad 💻  and totally a student community person who was fascinated 😮  by developer tools and DevOps when I first got into computer programming. My journey started from hardware engineering to developing web applications and backend scripts. Back then, I was learning React.js and developing projects. In hosting webpages using GitHub Pages, I got a glimpse of GitHub Actions. Being curious, I read the YAML file under the workflow directory and the script for CI/CD. Everything started making real sense to me. That moment made me learn more about CI/CD and workflows and how GitHub Actions can serve us. Keep reading to learn how you can use GitHub Actions to automate your workflows.

![Screenshot 2022-10-17 at 5 46 55 PM](https://user-images.githubusercontent.com/95834130/196174949-9a5a8b1f-477c-435d-af50-73d07358a036.png)

## Lights, camera, and ACTIONS! 📸

Let’s start with the obvious question - What are GitHub Actions? It is a simple continuous integration and continuous delivery (CI/CD) platform used to automate developer workflows.

GitHub Actions uses YAML syntax to define the workflow. Each workflow is stored as a separate YAML file in your code repository, in a directory named .github/workflows.

To put it simply- It’s an action or collection of actions that gets triggered when an event has occurred.

Wait! Then what about CI/CD? Actually, CI/CD (Continuous Integration/Continuous Development) is one of the many workflows that we can automate using GitHub Actions.

## What is a workflow?

A workflow is configured to begin once a specific trigger occurs, like when someone opens a pull request. It is an automated process that will run one or more jobs. For more information, see [Triggering a workflow](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow). A workflow typically consists of the following components:

- Jobs: a job represents a collection of 'steps' that executes on the same runner
- Steps: each ‘step’ executes logic to achieve some goal
- Runner: a virtual machine that runs your jobs

## Understanding how a workflow works ❓

Now that we know the basics. ✔️  Let’s get started! To keep things super simple, let us consider a real-life situation: collecting milk from the doorstep after it is delivered by a milkman.
```
name: Milk Collection
on: [when kept on door by milkman]  
jobs:  
    milkCollection:    
    runs-on: my-home   
    steps:      
     - uses: actions/checkout@v2      
     - name: milk-collector        
       uses: milk/action-KeepItInTheFridge@v3
```

Here, the Milk Collection named workflow will be executed when milk will be kept on the door by the milkman.

Once the event gets triggered, making it follow the jobs specified. In this case, there is one job `specified: milkCollection`.

Note: There can be multiple jobs which can follow previous workflows,  for example, `milkPayment` for paying the bill, or milkMessageAlert which will notify the delivery of milk:

`milkCollection` will run once we specify the environment under which it should run. Otherwise, it's quite possible it might start working for my neighbor's home, so precisely specifying my home as the environment.

Specifying the environment helps the later steps to be executed as per the requirement, allowing the job to finish properly. 

Specifying the order of events. Several `steps` should be considered before starting any of the work. For instance, in order to collect milk from our doorstep, we need to leave the current room we’re in, put on slippers, move towards the gate, and then open it.

![Screenshot 2022-10-17 at 5 47 54 PM](https://user-images.githubusercontent.com/95834130/196175154-21559a8d-4ff1-4758-9a7f-a919d9ea70e8.png)

Source: [Dreamstime Stock Images](https://www.dreamstime.com/stock-photos)

Once we’ve reached the gate and are able to collect the milk, we can now perform the action of placing it into the fridge.

BOOOMMM 💥!! 

![Screenshot 2022-10-17 at 5 48 54 PM](https://user-images.githubusercontent.com/95834130/196175311-9f867382-b91d-4ea7-b595-c250e627233f.png)

Source: [Insider](https://www.insider.com/)

Before moving ahead, you should know how and when we use [Events](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions#events)!

An event is any activity that happens in a GitHub repository and triggers a workflow run. It can include opening an issue, creating a pull request, making a commit, merging a PR, and/or joining a contributor. For instance, in our use case above ➡️ `on: [when kept on door by milkman]` is an event that triggers the workflow(process of delivering the milk). Here’s [a complete list of events](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows) that triggers workflows. 

### How can GitHub Actions help automate this workflow? 5 simple steps to set up a workflow
In order to automate your workflow and add the Actions template, let's build pipelines and workflows in our repository to have a practical demonstration. The key is to build breakpoints in a workflow with dependencies with conditional inputs.

**Situation**: We have an open-source repository and we wish to greet every new contributor.

![Screenshot 2022-10-17 at 5 49 59 PM](https://user-images.githubusercontent.com/95834130/196175499-eb720966-c0e9-404e-9bac-fdf5200194a6.png)

We are all set with a repository, ready for open-source contribution. People can fork it to contribute, but to enable that, we will set up a workflow into the YAML file which will do the required actions.

For setting up a workflow file .yml extension, follow these steps —

### 1️⃣ Step 1: Click on the Actions tab, which is present in the repo.

![Screenshot 2022-10-17 at 5 50 52 PM](https://user-images.githubusercontent.com/95834130/196175654-11331b9f-503c-482c-8b28-4670a9be0e64.png)

### 2️⃣ Step 2: Click on ‘set up a workflow yourself’.

![Screenshot 2022-10-17 at 5 51 36 PM](https://user-images.githubusercontent.com/95834130/196175772-e7e6b77e-8c3a-472e-9fdf-46393c3ce96a.png)

Note: GitHub will automatically create a `.github/workflows directory` for handling the YAML files.

```
`name: Greetings
on: [pull_request_target, issues]
jobs:
  greeting:
    runs-on: ubuntu-latest
    permissions:
      issues: write
      pull-requests: write
    steps:
    - uses: actions/first-interaction@v1
      with:
        repo-token: ${{ secrets.GITHUB_TOKEN }}
        issue-message: "You can be a problem solver! Go for it."
        pr-message: "Congo for your PR! Your PR may create an  
        Impact."`
```

You can use this workflow for your own setup. Just remember that in order to trigger a workflow, a pull_request_target or issues event should take place.

Note: `actions/first-interaction@v1` is being used. ‘Actions’ has its document which is a script, helping it quickly write a script in your workflow that uses the GitHub API and the workflow run context. Read more about it here. 

### 3️⃣ Step 3: Make it available to the public and ready for a Pull Request.

![Screenshot 2022-10-17 at 5 35 56 PM](https://user-images.githubusercontent.com/95834130/196172923-4dac75a9-eff1-4134-a932-e88896be4029.png)

The picture shows how a contributor is creating its pull request into the repository ‘actions-opensource’ where it will be triggering the workflow. 

As soon as workflow is triggered, actions will be checking the required dependencies.  

### 4️⃣ Step 4: GitHub-actions bot in action 🎊

![Screenshot 2022-10-17 at 5 36 37 PM](https://user-images.githubusercontent.com/95834130/196173092-33c33aa5-552b-4ba2-94e7-7f99ea277235.png)

After creating a PR successfully, the GitHub-actions bot will greet them with a welcome message since it is the very first Pull Request from the contributor.

The GitHub-actions bot can also be customized into any other greeting message, for instance when someone creates their first issue in this repo.

### 5️⃣ Step 5: The Pull Request gets accepted or updated based on the solution.
The final step is the acceptance of the Pull Request.

Hurray! 🎊 You’ve got the usage of workflow precisely using GitHub Actions. You have learnt how a workflow is set up and its structure. Keeping a note for this blog, the workflow was triggered by a pull request made into a repository. The triggering process will also be observed when an issue is raised in the repository. 

Some common issues that can be faced while or after setting up the workflows:

- "Resource not accessible by integration" - This happens when the contributor is not having relevant permission. To solve this, make sure your repository is having the read and write permission.
- GitHub Actions has encountered an internal error when running your job - This might happen due to technical issues from the GitHub outage. You can check the status on githubstatus.com or you can re-verify the configuration file of the environment being used.
- Action failed with error: `workload_identity_provider` or `credentials_json` - This simply happens when `env.SERVICE_JSON` is defined in the workflow for JSON and the fork tries to inject secrets using pull requests.

Hoping that you can now automate your workflows using GitHub Actions and enjoy the real sense of the CI/CD pipeline. 🚀

## TL;DR:

- CI/CD is a set of best practices that application development teams use to deliver code changes more frequently and reliably
- GitHub Actions is a CI/CD platform that offers an expressive configuration while supporting many community-driven third-party tools, all for free and directly integrated with where your code likely already exists, on the largest code repository platform available today.
- With GitHub Actions, you can trigger CI/CD workflows and pipelines of webhooks.There are a bunch of tools available to configure and execute the workflows such as ‘[circleci](https://circleci.com/)’ or ‘[Jenkins](https://www.jenkins.io/)’.

### References:

1. [Understanding GitHub Actions](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions)
2. [About workflows](https://docs.github.com/en/actions/using-workflows/about-workflows) 
3. [About CI/CD](https://www.redhat.com/en/topics/devops/what-cicd-pipeline)
